### PR TITLE
core: Add additional properties recursively for structured output(openai) when nested object is generated

### PR DESCRIPTION
**Pull request by fufufukakaka**
Hello. 
First of all, thank you for maintaining such a great project.

## Description
In https://github.com/langchain-ai/langchain/pull/25123, support for structured_output is added. However, `"additionalProperties": false` needs to be included at all levels when a nested object is generated.

error from current code: https://gist.github.com/fufufukakaka/e9b475300e6934853d119428e390f204
```
BadRequestError: Error code: 400 - {'error': {'message': "Invalid schema for response_format 'JokeWithEvaluation': In context=('properties', 'self_evaluation'), 'additionalProperties' is required to be supplied and to be false", 'type': 'invalid_request_error', 'param': 'response_format', 'code': None}}
```

Reference: [Introducing Structured Outputs in the API](https://openai.com/index/introducing-structured-outputs-in-the-api/)

```json
{
  "model": "gpt-4o-2024-08-06",
  "messages": [
    {
      "role": "system",
      "content": "You are a helpful math tutor."
    },
    {
      "role": "user",
      "content": "solve 8x + 31 = 2"
    }
  ],
  "response_format": {
    "type": "json_schema",
    "json_schema": {
      "name": "math_response",
      "strict": true,
      "schema": {
        "type": "object",
        "properties": {
          "steps": {
            "type": "array",
            "items": {
              "type": "object",
              "properties": {
                "explanation": {
                  "type": "string"
                },
                "output": {
                  "type": "string"
                }
              },
              "required": ["explanation", "output"],
              "additionalProperties": false
            }
          },
          "final_answer": {
            "type": "string"
          }
        },
        "required": ["steps", "final_answer"],
        "additionalProperties": false
      }
    }
  }
}
```

In the current code, `"additionalProperties": false` is only added at the last level. 
This PR introduces the `_add_additional_properties_key` function, which recursively adds `"additionalProperties": false` to the entire JSON schema for the request.

Twitter handle: `@fukkaa1225`

Thank you!

### CODE REVIEW

File: libs/partners/openai/tests/integration_tests/chat_models/test_base.py [hunk 1]
```diff
@@ -865,3 +865,45 @@ class InvalidJoke(BaseModelProper):
...
+    [("gpt-4o-2024-08-06", "json_schema", None)],
+)
+def test_nested_structured_output_strict(
+    model: str,
+    method: Literal["json_schema"],
+    strict: Optional[bool],
+) -> None:
+    """Test to verify structured output with strict=True for nested object."""
+
+    from typing import TypedDict
```
Comment (fufufukakaka) [inline libs/partners/openai/tests/integration_tests/chat_models/test_base.py:881]: now, pydantic outputs json schema includes 'allOf', but openai don't support.
pydantic PR → https://github.com/pydantic/pydantic/pull/10029

```
In [21]: result = chat.invoke("Tell me a joke about cats.")
---------------------------------------------------------------------------
BadRequestError                           Traceback (most recent call last)
Cell In[21], line 1
----> 1 result = chat.invoke("Tell me a joke about cats.")

File ~/Library/Caches/pypoetry/virtualenvs/icebreak-generator-SqI8j0ct-py3.12/lib/python3.12/site-packages/langchain_core/runnables/base.py:2876, in RunnableSequence.invoke(self, input, config, **kwargs)
   2874 context.run(_set_config_context, config)
   2875 if i == 0:
-> 2876     input = context.run(step.invoke, input, config, **kwargs)
   2877 else:
   2878     input = context.run(step.invoke, input, config)

File ~/Library/Caches/pypoetry/virtualenvs/icebreak-generator-SqI8j0ct-py3.12/lib/python3.12/site-packages/langchain_core/runnables/base.py:5094, in RunnableBindingBase.invoke(self, input, config, **kwargs)
   5088 def invoke(
   5089     self,
   5090     input: Input,
   5091     config: Optional[RunnableConfig] = None,
   5092     **kwargs: Optional[Any],
   5093 ) -> Output:
-> 5094     return self.bound.invoke(
   5095         input,
   5096         self._merge_configs(config),
   5097         **{**self.kwargs, **kwargs},
   5098     )

File ~/Library/Caches/pypoetry/virtualenvs/icebreak-generator-SqI8j0ct-py3.12/lib/python3.12/site-packages/langchain_core/language_models/chat_models.py:284, in BaseChatModel.invoke(self, input, config, stop, **kwargs)
    273 def invoke(
    274     self,
    275     input: LanguageModelInput,
   (...)
    279     **kwargs: Any,
    280 ) -> BaseMessage:
    281     config = ensure_config(config)
    282     return cast(
    283         ChatGeneration,
--> 284         self.generate_prompt(
    285             [self._convert_input(input)],
    286             stop=stop,
    287             callbacks=config.get("callbacks"),
    288             tags=config.get("tags"),
    289             metadata=config.get("metadata"),
    290             run_name=config.get("run_name"),
    291             run_id=config.pop("run_id", None),
    292             **kwargs,
    293         ).generations[0][0],
    294     ).message

File ~/Library/Caches/pypoetry/virtualenvs/icebreak-generator-SqI8j0ct-py3.12/lib/python3.12/site-packages/langchain_core/language_models/chat_models.py:756, in BaseChatModel.generate_prompt(self, prompts, stop, callbacks, **kwargs)
    748 def generate_prompt(
    749     self,
    750     prompts: List[PromptValue],
   (...)
    753     **kwargs: Any,
    754 ) -> LLMResult:
    755     prompt_messages = [p.to_messages() for p in prompts]
--> 756     return self.generate(prompt_messages, stop=stop, callbacks=callbacks, **kwargs)

File ~/Library/Caches/pypoetry/virtualenvs/icebreak-generator-SqI8j0ct-py3.12/lib/python3.12/site-packages/langchain_core/language_models/chat_models.py:613, in BaseChatModel.generate(self, messages, stop, callbacks, tags, metadata, run_name, run_id, **kwargs)
    611         if run_managers:
    612             run_managers[i].on_llm_error(e, response=LLMResult(generations=[]))
--> 613         raise e
    614 flattened_outputs = [
    615     LLMResult(generations=[res.generations], llm_output=res.llm_output)  # type: ignore[list-item]
    616     for res in results
    617 ]
    618 llm_output = self._combine_llm_outputs([res.llm_output for res in results])

File ~/Library/Caches/pypoetry/virtualenvs/icebreak-generator-SqI8j0ct-py3.12/lib/python3.12/site-packages/langchain_core/language_models/chat_models.py:603, in BaseChatModel.generate(self, messages, stop, callbacks, tags, metadata, run_name, run_id, **kwargs)
    600 for i, m in enumerate(messages):
    601     try:
    602         results.append(
--> 603             self._generate_with_cache(
    604                 m,
    605                 stop=stop,
    606                 run_manager=run_managers[i] if run_managers else None,
    607                 **kwargs,
    608             )
    609         )
    610     except BaseException as e:
    611         if run_managers:

File ~/Library/Caches/pypoetry/virtualenvs/icebreak-generator-SqI8j0ct-py3.12/lib/python3.12/site-packages/langchain_core/language_models/chat_models.py:825, in BaseChatModel._generate_with_cache(self, messages, stop, run_manager, **kwargs)
    823 else:
    824     if inspect.signature(self._generate).parameters.get("run_manager"):
--> 825         result = self._generate(
    826             messages, stop=stop, run_manager=run_manager, **kwargs
    827         )
    828     else:
    829         result = self._generate(messages, stop=stop, **kwargs)

File ~/Library/Caches/pypoetry/virtualenvs/icebreak-generator-SqI8j0ct-py3.12/lib/python3.12/site-packages/langchain_openai/chat_models/base.py:629, in BaseChatOpenAI._generate(self, messages, stop, run_manager, **kwargs)
    624         warnings.warn(
    625             "Cannot currently include response headers when response_format is "
    626             "specified."
    627         )
    628     payload.pop("stream")
--> 629     response = self.root_client.beta.chat.completions.parse(**payload)
    630 elif self.include_response_headers:
    631     raw_response = self.client.with_raw_response.create(**payload)

File ~/Library/Caches/pypoetry/virtualenvs/icebreak-generator-SqI8j0ct-py3.12/lib/python3.12/site-packages/openai/resources/beta/chat/completions.py:112, in Completions.parse(self, messages, model, response_format, frequency_penalty, function_call, functions, logit_bias, logprobs, max_tokens, n, parallel_tool_calls, presence_penalty, seed, service_tier, stop, stream_options, temperature, tool_choice, tools, top_logprobs, top_p, user, extra_headers, extra_query, extra_body, timeout)
    105 _validate_input_tools(tools)
    107 extra_headers = {
    108     "X-Stainless-Helper-Method": "beta.chat.completions.parse",
    109     **(extra_headers or {}),
    110 }
--> 112 raw_completion = self._client.chat.completions.create(
    113     messages=messages,
    114     model=model,
    115     response_format=_type_to_response_format(response_format),
    116     frequency_penalty=frequency_penalty,
    117     function_call=function_call,
    118     functions=functions,
    119     logit_bias=logit_bias,
    120     logprobs=logprobs,
    121     max_tokens=max_tokens,
    122     n=n,
    123     parallel_tool_calls=parallel_tool_calls,
    124     presence_penalty=presence_penalty,
    125     seed=seed,
    126     service_tier=service_tier,
    127     stop=stop,
    128     stream_options=stream_options,
    129     temperature=temperature,
    130     tool_choice=tool_choice,
    131     tools=tools,
    132     top_logprobs=top_logprobs,
    133     top_p=top_p,
    134     user=user,
    135     extra_headers=extra_headers,
    136     extra_query=extra_query,
    137     extra_body=extra_body,
    138     timeout=timeout,
    139 )
    140 return _parse_chat_completion(
    141     response_format=response_format,
    142     chat_completion=raw_completion,
    143     input_tools=tools,
    144 )

File ~/Library/Caches/pypoetry/virtualenvs/icebreak-generator-SqI8j0ct-py3.12/lib/python3.12/site-packages/openai/_utils/_utils.py:274, in required_args.<locals>.inner.<locals>.wrapper(*args, **kwargs)
    272             msg = f"Missing required argument: {quote(missing[0])}"
    273     raise TypeError(msg)
--> 274 return func(*args, **kwargs)

File ~/Library/Caches/pypoetry/virtualenvs/icebreak-generator-SqI8j0ct-py3.12/lib/python3.12/site-packages/openai/resources/chat/completions.py:650, in Completions.create(self, messages, model, frequency_penalty, function_call, functions, logit_bias, logprobs, max_tokens, n, parallel_tool_calls, presence_penalty, response_format, seed, service_tier, stop, stream, stream_options, temperature, tool_choice, tools, top_logprobs, top_p, user, extra_headers, extra_query, extra_body, timeout)
    616 @required_args(["messages", "model"], ["messages", "model", "stream"])
    617 def create(
    618     self,
   (...)
    648     timeout: float | httpx.Timeout | None | NotGiven = NOT_GIVEN,
    649 ) -> ChatCompletion | Stream[ChatCompletionChunk]:
--> 650     return self._post(
    651         "/chat/completions",
    652         body=maybe_transform(
    653             {
    654                 "messages": messages,
    655                 "model": model,
    656                 "frequency_penalty": frequency_penalty,
    657                 "function_call": function_call,
    658                 "functions": functions,
    659                 "logit_bias": logit_bias,
    660                 "logprobs": logprobs,
    661                 "max_tokens": max_tokens,
    662                 "n": n,
    663                 "parallel_tool_calls": parallel_tool_calls,
    664                 "presence_penalty": presence_penalty,
    665                 "response_format": response_format,
    666                 "seed": seed,
    667                 "service_tier": service_tier,
    668                 "stop": stop,
    669                 "stream": stream,
    670                 "stream_options": stream_options,
    671                 "temperature": temperature,
    672                 "tool_choice": tool_choice,
    673                 "tools": tools,
    674                 "top_logprobs": top_logprobs,
    675                 "top_p": top_p,
    676                 "user": user,
    677             },
    678             completion_create_params.CompletionCreateParams,
    679         ),
    680         options=make_request_options(
    681             extra_headers=extra_headers, extra_query=extra_query, extra_body=extra_body, timeout=timeout
    682         ),
    683         cast_to=ChatCompletion,
    684         stream=stream or False,
    685         stream_cls=Stream[ChatCompletionChunk],
    686     )

File ~/Library/Caches/pypoetry/virtualenvs/icebreak-generator-SqI8j0ct-py3.12/lib/python3.12/site-packages/openai/_base_client.py:1259, in SyncAPIClient.post(self, path, cast_to, body, options, files, stream, stream_cls)
   1245 def post(
   1246     self,
   1247     path: str,
   (...)
   1254     stream_cls: type[_StreamT] | None = None,
   1255 ) -> ResponseT | _StreamT:
   1256     opts = FinalRequestOptions.construct(
   1257         method="post", url=path, json_data=body, files=to_httpx_files(files), **options
   1258     )
-> 1259     return cast(ResponseT, self.request(cast_to, opts, stream=stream, stream_cls=stream_cls))

File ~/Library/Caches/pypoetry/virtualenvs/icebreak-generator-SqI8j0ct-py3.12/lib/python3.12/site-packages/openai/_base_client.py:936, in SyncAPIClient.request(self, cast_to, options, remaining_retries, stream, stream_cls)
    927 def request(
    928     self,
    929     cast_to: Type[ResponseT],
   (...)
    934     stream_cls: type[_StreamT] | None = None,
    935 ) -> ResponseT | _StreamT:
--> 936     return self._request(
    937         cast_to=cast_to,
    938         options=options,
    939         stream=stream,
    940         stream_cls=stream_cls,
    941         remaining_retries=remaining_retries,
    942     )

File ~/Library/Caches/pypoetry/virtualenvs/icebreak-generator-SqI8j0ct-py3.12/lib/python3.12/site-packages/openai/_base_client.py:1040, in SyncAPIClient._request(self, cast_to, options, remaining_retries, stream, stream_cls)
   1037         err.response.read()
   1039     log.debug("Re-raising status error")
-> 1040     raise self._make_status_error_from_response(err.response) from None
   1042 return self._process_response(
   1043     cast_to=cast_to,
   1044     options=options,
   (...)
   1048     retries_taken=options.get_max_retries(self.max_retries) - retries,
   1049 )

BadRequestError: Error code: 400 - {'error': {'message': "Invalid schema for response_format 'JokeWithEvaluation': In context=('properties', 'self_evaluation'), 'allOf' is not permitted", 'type': 'invalid_request_error', 'param': 'response_format', 'code': None}}
```

At this point, I will only implement tests using TypedDict. 🙏

**Comment by sajal2692** [comment]
When is this expected to be merged? would be super useful

File: libs/core/langchain_core/utils/function_calling.py
Comment (baskaryan) [inline libs/core/langchain_core/utils/function_calling.py:1]: could we add unit tests for this functionality instead of integration tests?

**Review by baskaryan — APPROVED** [review]

**Comment by baskaryan** [comment]
thank you @fufufukakaka! will be out in langchain-core 0.2.35, which will be released later today

**Comment by fufufukakaka** [comment]
thanks!

**Pull request merged**